src/util/misc/matrix.ts

```diff
@@ -0,0 +1,209 @@
+import { cos } from "./cos";
+import { sin } from "./sin";
+import { degreesToRadians } from "./radiansDegreesConversion";
+import { iMatrix, PiBy180 } from "../../constants";
+import { IPoint, Point } from "../../point.class";
+import { TDegree, TMat2D } from "../../typedefs";
+
+type TRotateMatrixArgs = {
+  angle?: TDegree;
+}
+
+type TTranslateMatrixArgs = {
+  translateX?: number;
+  translateY?: number;
+}
+
+type TScaleMatrixArgs = {
+  scaleX?: number;
+  scaleY?: number;
+  flipX?: boolean;
+  flipY?: boolean;
+  skewX?: TDegree;
+  skewY?: TDegree;
+}
+
+type TComposeMatrixArgs = TTranslateMatrixArgs & TRotateMatrixArgs & TScaleMatrixArgs;
+/**
+ * Apply transform t to point p
+ * @static
+ * @memberOf fabric.util
+ * @param  {Point | IPoint} p The point to transform
+ * @param  {Array} t The transform
+ * @param  {Boolean} [ignoreOffset] Indicates that the offset should not be applied
+ * @return {Point} The transformed point
+ */
+export const transformPoint = (p: Point | IPoint, t: TMat2D, ignoreOffset: boolean): Point => {
+  if (ignoreOffset) {
+    return new Point(
+      t[0] * p.x + t[2] * p.y,
+      t[1] * p.x + t[3] * p.y
+    );
+  }
+  return new Point(
+    t[0] * p.x + t[2] * p.y + t[4],
+    t[1] * p.x + t[3] * p.y + t[5]
+  );
+};
+
+/**
+ * Invert transformation t
+ * @static
+ * @memberOf fabric.util
+ * @param {Array} t The transform
+ * @return {Array} The inverted transform
+ */
+export const invertTransform = (t: TMat2D): TMat2D => {
+  const a = 1 / (t[0] * t[3] - t[1] * t[2]),
+        r = [a * t[3], -a * t[1], -a * t[2], a * t[0], 0, 0] as TMat2D,
+        { x, y } = transformPoint({ x: t[4], y: t[5] }, r, true);
```

Comment:
we should decide about the fate of Point and IPoint

Reply:
The reason why i didn't overuse the point years ago is because i thought a plain object would be more performant than a function or a class. But i think this is not true.
I never benchmarked it, and i struggle to understand why fabric starts to cycle through an handful of hundreds objects.

I like that a Point can be a Point or its basic interface because it allows you to reuse the data you have without transforming it.

Reply:
I think we need to aim for flexibility but tend to use Point in fabric and not a simple object.
I don't think there a real diff in perf.
All major repos use classes so it has to be a good idea and the browser as well